### ISSUE TITLE
Make insert-mode `move-dot-*` and `kill-*` functions "pure"

### ISSUE THREAD

**Issue by kwshi** [elves/elvish#722]
Not a bug, but an implementation suggestion.

Currently, insert-mode navigation functions (i.e. `move-*` and `kill-*`) modify the editor state in-place.  This is not inherently bad, and works perfectly fine.  However, I argue that it is more desirable for these functions to not modify the state in-place but simply return new editor state variables (namely `ed.buffer` and `ed.dot`).

---

Here's the proposed change, in a little bit more detail.  Currently, `move-dot-*` functions directly modify the editor state:
```go
func (ed *editor) moveDotSomewhere() {
  // mutates `ed`
  ed.dot = // some computation...
}
```
Notice that the `move-dot-*` family of functions should depend *only* on the current buffer input and cursor position and should not at all touch any other portion of the editor state (e.g. `notifications`, `tips`, `styling`), much less the rest of the editor object.  The only thing they should modify is the cursor position, or `ed.dot`.  Thus we make these functions pure, and restrict their access to only the variables they should use:
```go
func moveDotSomewhere(buffer string, dot int) int {
  newDot := // some computation...
  return newDot
}
```

Analogously, the `kill-*` functions, currently defined as
```go
func (ed *editor) killSomething() {
  ed.buffer = // some modified buffer contents...
  ed.dot = // new dot position...
}
```
would be revised to only take the current buffer contents and cursor position, and return new buffer contents and cursor position.  They should have nothing to do with any other part of the editor.
```go
func killSomething(buffer string, dot int) (string, int) {
  newBuffer := // some modified buffer contents...
  newDot := // new dot position...
  return newBuffer, newDot
}
```

Because they are now pure functions, we do need a *little* bit of extra work to apply the effects.  Instead of 
```go
func initCoreFns(ed *editor, ns eval.Ns) {
  ns.AddBuiltinFns("edit:", map[string]interface{}{
    // ...
    "kill-something": killSomething,
    // ...
    "move-dot-somewhere": moveDotSomewhere,
    // ...
  })
}
```
we will have to do something like
```go
func initCoreFns(ed *editor, ns eval.Ns) {
  ns.AddBuiltinFns("edit:", map[string]interface{}{
    // ...
    "kill-something": ed.applyKill(killSomething),
    // ...
    "move-dot-somewhere": ed.applyMove(moveDotSomewhere),
    // ...
  })
}

func (ed *editor) applyKill(fn func(string, int) (string, int)) {
  return func() {
    ed.buffer, ed.dot = fn(ed.buffer, ed.dot)
  }
}

func (ed *editor) applyMove(fn func(string, int) int) {
  return func() {
    ed.dot = fn(ed.buffer, ed.dot)
  }
}
```
Of course, we can do away with some of the repetitive `applyKill()` and `applyMove()` boilerplate by defining some batch function factory, such as
```go
func convertMoveFns(fns map[string]func(string, int) int) map[string]interface{} {
  newFns := make(map[string]interface{})
  for name, fn := range fns {
    newFns[name] = applyMove(fn)
  }
  return newFns
}

func convertKillFns(fns map[string]func(string, int) (string, int)) map[string]interface{} {
  newFns := make(map[string]interface{})
  for name, fn := range fns {
    newFns[name] = applyKill(fn)
  }
  return newFns
}
```
whose outputs can be passed to `ns.AddBuiltinFns` to apply the converted pure functions in batch.

---

Here are some reasons I consider this change favorable:

- It adheres to the whole functional programming "philosophy"?  Immutability and all?  Someone might have more to say about this.  I mention it only for aesthetic considerations.
- It makes debugging and unit-testing much, *much* easier.  By isolating the actual input and output of the move and kill functions, we make it much more straightforward to write test cases.  This way, we can systematically and predictably check that implementations of these movement functions behave as intended, without having to compile and hand-test the behavior interactively, or go through some hassle of constructing an entire `editor` object for testing.
- There is added safety (type safety, too!), because, again, since the input and output of the functions are isolated, we absolutely ensure that the movement and kill functions do not access any other part of the editor and unpredictably introduce some error.  It is of course relatively unlikely that someone who knows what they're doing *accidentally* touch anything other than the buffer and the dot position, but this sort of safety is always a good idea, and arguably makes the code more modularized, more contained, etc.
- This approach also facilitates some Elvish-scriptable API (eventually, when we have a well-established API) for cursor movement, since the `applyMove` and `applyKill` functions could be exported into the `edit:` namespace, and then users would be able to define their own move and kill *pure functions*, with which they can extend/modify cursor movement behavior in Elvish, via something like
  ```elvish
  fn my-own-move [buffer dot] {
    new-dot = // some computation
    put $new-dot
  }
  edit:insert:binding[Ctrl-k] = edit:apply-move $my-own-move~
  ```
  (I concede: I do not yet understand the mechanics of Elvish very well yet, so the above example may very well be wrong/broken, but I believe it's conceptually feasible and more or less conveys the right idea.)

**Comment by kwshi** [comment]
There is yet another (relatively minor) side-benefit of this idea, which is that we would be able to reuse some of the computation for free.

Currently, for example, the `moveDotLeftWord` and `killWordLeft` functions effectively perform the same computation, but do it independently.  A good refactoring idea would be to put the computation of the index in a separate function (something like `computeLeftWordIndex(buffer string, dot int) int`), and then to use that function in both `moveDotLeftWord` and `killWordLeft`.

However, if we refactor these functions to be pure, we might notice that `computeLeftWordIndex` is effectively the same thing as `moveDotLeftWord`, and we would be able to trivially reuse the logic with something like
```go
func killWordLeft(buffer string, dot int) (string, int) {
  index := moveDotLeftWord(buffer, dot)
  return buffer[:index] + buffer[dot:], index
}
```

In fact, we can even automate creation of `kill` functions from `move` functions:
```go
func killMove(move func(string, int) int) func(string, int) (string, int) {
  return func(buffer string, dot int) (string, int) {
    index := move(buffer, dot)
    if index < dot { // delete left
      return buffer[:index] + buffer[dot:], index
    } else { // delete right
      return buffer[:dot] + buffer[index:], dot
    }
  }
}
```

This ends up saving quite a bit of boilerplate and repetition, even.

**Comment by kwshi** [comment]
Yes.  For most cursor-based operations, `kill` can be derived directly from `move`.

However, now that I think about it, maybe it's also worth considering a most-general `modifyBuffer(string, int) (string, int)`, which would (in terms of API) allow fully flexible buffer manipulation in a way that `move`-derived killing does not (perhaps I'd like to lowercase the entire buffer? etc.).

**Comment by kwshi** [comment]
Resolved by #721.
